File: demo_onto/ili_sandbox/populate_ili.pl
# populate_ili.pl May 2010
# Mike Conway, DBMI Pittsburgh

# Perl script populates the ILI ontology
# SVN Version Info:
# $Id$

use RDF::Helper;

my $rdf = RDF::Helper->new();
$rdf->include_rdfxml(filename => "ili.owl");

while (<>) {
    if ($. == 1) {next}
    chomp;
    s/"//g;
    my @field = split(/\t/, $_);
    if ($. == 60) {last}

    
    my $subconcept = $field[16];  $subconcept = trim($subconcept);
    my $relation   = $field[17];  $relation = trim($relation);    
    my $umls       = $field[18];  $umls = trim($umls);
    my $diagnosis  = $field[19];  $diagnosis = trim($diagnosis);
    my $syndrome   = $field[20];  $syndrome = trim($syndrome);
    my $phys_find  = $field[21];  $syndrome = trim($phys_find);
    my $symptom    = $field[22];  $symptom = trim($symptom);
    my $sign       = $field[23];  $sign = trim($sign);

    #lowercase first letter of subconcept
    $subconcept = lcfirst($subconcept);

    # ADD DIAGNOSIS INSTANCES
    if (($diagnosis == 1) || ($syndrome == 1)) {
        $rdf->assert_resource("http://www.owl-ontologies.com/Ontology1274288723.owl#$subconcept","http://www.w3.org/1999/02/22-rdf-syntax-ns#type","http://www.owl-ontologies.com/Ontology1274288723.owl#Disease");
    }

    # ADD SIGNS AND PHYSICAL FINDING INSTANCES
    if (($sign == 1) || ($physical_finding == 1)) {
       $rdf->assert_resource("http://www.owl-ontologies.com/Ontology1274288723.owl#$subconcept","http://www.w3.org/1999/02/22-rdf-syntax-ns#type","http://www.owl-ontologies.com/Ontology1274288723.owl#Finding");
   }

    # ADD SYMPTOM INSTANCES
    if ($symptom == 1) {
        $rdf->assert_resource("http://www.owl-ontologies.com/Ontology1274288723.owl#$subconcept","http://www.w3.org/1999/02/22-rdf-syntax-ns#type","http://www.owl-ontologies.com/Ontology1274288723.owl#Symptom");
    }

    # ADD LABEL
    my $label_subconcept = $subconcept;
    $label_subconcept =~ s/([A-Z])/ $1/g;
    $label_subconcept = lc($label_subconcept);
    $label_subconcept = trim($label_subconcept);
    my $label_literal = $rdf->new_literal($label_subconcept, "en");
    $rdf->assert_literal("http://www.owl-ontologies.com/Ontology1274288723.owl#$subconcept","http://www.owl-ontologies.com/Ontology1274288723.owl#label", $label_literal);
    print "$label_subconcept" . "\n";
    
}



$rdf->serialize(filename => 'out.owl', format => 'rdfxml');


sub trim() {
	my $string = shift;
	$string =~ s/^\s+//;
	$string =~ s/\s+$//;
	return $string;
}

File: demo_onto/ili_sandbox/populate_ili.py
#!/usr/bin/python


# 19th May 2010
# populate_ili.py
# Mike Conway, University of Pittsburgh, DBMI

# SVN versioning info:
# $Id$

# This python script reads from a spreadsheet containing terms, relations, links,
# etc.  to populate a skeleton OWL file produced by Protege.  Currently we are
# only interested in ILI (Influenza like Illness).  The data was produced by
# Mike Conway and John Dowling (infectious disease physician)

import fileinput
import sys
import re
import pdb
from rdflib import *
import warnings


#load up local owl file containing relations and classes
graph = ConjunctiveGraph()
graph.commit()
graph.parse("ili.owl")




loop_counter = 0

for record in fileinput.input():
    #Skip headers line
    #Process those records only relevant to ILI (lns 2->59)
    if loop_counter == 0:
        loop_counter = loop_counter + 1
        continue
    if loop_counter == 60:
        break
    
    # First, remove all double quotation marks introduced
    # by the Excel "save to tab format"
    record = re.sub("\"", "", record)
    #pdb.set_trace()

    # FORMAT INPUT RECORD
    field = record.split('\t')
    subconcept = field[16]; subconcept.strip(); subconcept = unicode(subconcept)
    relation   = field[17]; relation.strip()
    umls       = field[18]; umls.strip()
    diagnosis  = field[19]; diagnosis.strip()
    syndrome   = field[20]; syndrome.strip()
    phys_find  = field[21]; phys_find.strip()
    symptom    = field[22]; symptom.strip()
    sign       = field[23]; sign.strip()

  
    
    if diagnosis == "1":
        graph.add((u"http://www.owl-ontologies.com/Ontology1274288723.owl#" + subconcept,u"http://www.w3.org/1999/02/22-rdf-syntax-ns#type", u"http://www.owl-ontologies.com/Ontology1274288723.owl#Disease"))
    

  
    
   
    loop_counter = loop_counter + 1

    #ADD subroutine to Add subconcepts (first diagnosis, then others)
    #                      UMLS code
    #                      keywords
    #                      regular expressions

graph.serialize(destination="output.owl", format="xml")

